c10/macros/Macros.h

```diff
@@ -331,8 +331,12 @@ constexpr uint32_t CUDA_THREADS_PER_BLOCK_FALLBACK = 256;
 // CUDA_KERNEL_ASSERT checks the assertion
 // even when NDEBUG is defined. This is useful for important assertions in CUDA
 // code that would otherwise be suppressed when building Release.
+//
+// Declaration of __assert_fail() does not match that of ROCm (even >= 4.1),
+// so CUDA_KERNEL_ASSERT() is disabled for HIP code when NDEBUG is set.
 #if defined(__ANDROID__) || defined(__APPLE__) || defined(__XROS__) || \
-    (defined(USE_ROCM) && ROCM_VERSION < 40100)
+    (defined(USE_ROCM) && ROCM_VERSION < 40100) ||                     \
+    (defined(NDEBUG) && (defined(__HIP_ARCH__) || defined(__HIP__)))
```

Comment:
Honestly this feels like HIP is just not ready for the NDEBUG flag. How about just change it to USE_ROCM, and when it's ready we'll turn it back on? What do you think @jeffdaily @jithunnair-amd

Reply:
The PR that enabled kernel asserts for ROCm (#49624) should be partially reverted, yes.  Partial, meaning update these lines here to disable kernel asserts for ROCm, without version condition.  But the original PR made improvements to the unit tests; the tests should be skipped now but keep the other improvements.  Unfortunately, the unit tests that would exercise the kernel asserts only execute for the "slow" category of tests, and ROCm does not have a GitHub Action for "slow" at this time.  Further, perhaps related to this current PR and the incorrect function declaration, we're seeing other issues with kernel asserts in pytorch starting in ROCm 4.5.2 and through 5.0.  The feature should be disabled until we can fully revisit.

Note that we used to set NDEBUG flag for all ROCm builds until that PR.

Reply:
Okay, then I will disable `CUDA_KERNEL_ASSERT()` for ROCm. Please let me update the PR.

Reply:
Thanks for your comments. I updated the commit. Now it passes all tests including the ROCm ones (https://github.com/pytorch/pytorch/runs/5363652718?check_suite_focus=true).

(I'd appreciate any comments/reviews from you: @jeffdaily, @jithunnair-amd)